// @milkdown/crepe/lib/cjs/index-YoPS8A3_.js
'use strict';

var linkTooltip = require('@milkdown/kit/component/link-tooltip');
var todoList = require('./todoList-Qb2JsRGC.js');
require('atomico');

const defineFeature = (editor, config) => {
  editor.config(linkTooltip.configureLinkTooltip).config((ctx) => {
    ctx.update(linkTooltip.linkTooltipConfig.key, (prev) => {
      var _a, _b, _c, _d, _e, _f;
      return {
        ...prev,
        linkIcon: (_a = config == null ? void 0 : config.linkIcon) != null ? _a : () => todoList.copyIcon,
        editButton: (_b = config == null ? void 0 : config.editButton) != null ? _b : () => todoList.editIcon,
        removeButton: (_c = config == null ? void 0 : config.removeButton) != null ? _c : () => todoList.removeIcon,
        confirmButton: (_d = config == null ? void 0 : config.confirmButton) != null ? _d : () => todoList.confirmIcon,
        inputPlaceholder: (_e = config == null ? void 0 : config.inputPlaceholder) != null ? _e : "Paste link...",
        onCopyLink: (_f = config == null ? void 0 : config.onCopyLink) != null ? _f : () => {
        }
      };
    });
  }).use(linkTooltip.linkTooltipPlugin);
};

exports.defineFeature = defineFeature;
